### DISCUSSION TITLE
error: does not look like an OMNeT++ root directory

### DISCUSSION THREAD

**Discussion by JuliaSteiwer** [omnetpp/omnetpp#1069] [Q&A] [answered]
Hello! I had to reinstall OMNeT++ on my Mac (macOS Mojave 10.14.6 [can't update this]) and now when I use "source setenv", I get the following error message for version 6.0.2: `does not look like an OMNeT++ root directory`

When I try to use 6.0 or 6.0.1 instead, I get the message that I'm using the Linux or Windows version, even though I installed the macOS version of OMNeT++ right from the official website.

Does anyone perhaps have an idea on how to fix this issue? I updated rvm yesterday and tried, but ultimately failed, to update Ruby (still at ruby 2.3.7p456), but these are the only changes I did on my laptop.

Any help is greatly appreciated; thank you already for your help!

Sincerely,
Julia

**Comment by rhornig** [comment]
1. ruby has absolutely nothing to do with OMNeT++. It does not use it at all.
The error message is directly from the setenv:
```
if [ ! -f "$dir/configure.user" -o ! -f "$dir/include/omnetpp.h" -o ! -f "$dir/Version" ]; then
echo "Error: '$dir' does not look like an OMNeT++ root directory"
return 1
fi
```

It just checks for the presence of the above files. Check them for yourself manually. If they are not there, something is seriously broken.

**Reply by JuliaSteiwer** [reply]
Hello and thank you for your quick reply; my apologies for not replying sooner. 

The file `configure.user` exists, and its content looks good, too. Next, `/include/omnetpp.h` is there as well, and also looks fine. The "Version" file also exists, it only contains the line `omnetpp-6.0.2`. So these files are all there.

Maybe I made a mistake during installation? Since I already had OMNeT++ installed previously, I started again at step 4.7 of the installation manual. I downloaded the 6.0.2 version for macOS from the official page into my user folder, then I opened the zipped file using the command `tar zxvf omnetpp-6.0.2-macos-x86_64.tgz`. That unpacked everything normally. Next, I used the `cd omnetpp-6.0.2` command, which worked fine. Using `source setenv` next, I get the aforementioned error. If I try to run `./configure` or `make`, I get the error message that I have to use `source setenv` first - which doesn't work.

Does this additional information perhaps give you a better idea of what my problem could be? I asked my academic supervisor about it and he has never faced this issue before. He also suggested that it might be because of the files. I have attached the files as PDFs for your convenience.

Thank you!

[configure_user.pdf](https://github.com/omnetpp/omnetpp/files/13323651/configure_user.pdf) (this is the `configure.user` file)
[omnetpp_h.pdf](https://github.com/omnetpp/omnetpp/files/13323660/omnetpp_h.pdf) (this is the `include/omnetpp.h` file)

**Reply by rhornig** [reply]
What is the **exact** (and complete) output when you execute `source setenv` ? The script just checks for the presence of those files so their content is not important. But the directory name in the error message would be interesting.

To fail this test, you either have to miss one of those files or the script cannot figure out its own base directory. What shell you are running? bash or zsh ?

**Reply by JuliaSteiwer** [reply]
The full output looks like this:

```
Julias-MBP:~ juliasteiwer$ cd omnetpp-6.0.2/
Julias-MBP:omnetpp-6.0.2 juliasteiwer$ source setenv
Error: '/Users/juliasteiwer/omnetpp-6.0.2
Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.' does not look like an OMNeT++ root directory
Julias-MBP:omnetpp-6.0.2 juliasteiwer$ 
``` 

I just use the default, which - given that I'm on macOS 10.14.6 - should be bash.

**Reply by rhornig** [reply]
The fact that all version (6.0.0, 6.0.1 and 6.0.2) fail to detect an existing file tells that there is something wrong with your system. I'm not sure what, but that is global and not related to the omnet installation. Could be some permission issue, or some corrupted system utils.

If you `cd omnetpp-6.0.2` and then type `[ -f Version ]; echo $? ; ls -l Version` what is the exact output?

**Reply by JuliaSteiwer** [reply]
That'd be a massive problem … just weird since 6.0.1 used to work at one point. 
The exact output of that command is
```
0
-rw-r--r--@ 1 juliasteiwer  staff  14  6 Okt 22:38 Version
```
(which looks bad)

**Reply by rhornig** [reply]
No, the output looks absolutely OK. (except for the extended attribute that looks a bit suspicious) What does `xattr Version` say?

Also what is the output of `[ ! "configure.user" -o ! -f "include/omnetpp.h" -o ! -f "Version" ]; echo $?`

**Reply by JuliaSteiwer** [reply]
Using the command `xattr Version`, I get this output:
```
com.apple.lastuseddate#PS
com.apple.quarantine
```
When I use the other command, it just returns `1`.

**Reply by rhornig** [reply]
I'm lost here. The same test command that evaluates in the setenv script to TRUE, here evaluates to FALSE. This cannot happen, unless... I don't know.

Copy the `setenv` script to an other file with a different name and start to debug. Delete everything after line 23 and then add debug staments. echo the $dir variable. Try to pin down why the same expression evaluates differently when invoked in the setenv compared to executed it directly in the bash shell.

There is a slight chance that the quarantine extended attribute causing this. I do not have that attribute set on my files.

Just out of curiosity, how did you extract the archive? What program was used? (I'm asking this because anything downloaded from the internet would automatically set the quarantine bit on any extracted file).

**Reply by JuliaSteiwer** [reply]
Okay, that sounds like a major issue.

I tried debugging the `setenv` file in VSCode but it didn't let me, always complaining that the debug version in the JSON file wasn't right, and even when I set it to the right version (4.* or 5.*), it still complained about the exact same issue.

How would I unset this attribute?

Anyway, I completely deleted everything OMNeT++ from my laptop and rebooted my laptop as well. Normally, I used the `tar zxvf omnetpp-6.0-macos-x86_64.tgz` command to unpack the archive. This time I used the Finder to unpack but the issue with `source setenv` remains the same. For your convenience, I have attached a PDF version of my `setenv` file, maybe this helps?
[setenv.pdf](https://github.com/omnetpp/omnetpp/files/13343858/setenv.pdf)

Executing the previous commands you wrote, they return the same things as before except `xattr Version`, which only returns `com.apple.quarantine` this time.

**Reply by rhornig** [reply]
Sadly I don't have any more idea. When I said debugging I meant that you have an expression that works as expected ( `[ ! "configure.user" -o ! -f "include/omnetpp.h" -o ! -f "Version" ]; echo $?` returns 1). While the same command inside the setenv script returns 0 (because you got an error message on evaluating that expression). You should start modifiying/reducing the setenv script until you get the correct return value. That process MAY give you a hint, what is wrong.

**Reply by JuliaSteiwer** [reply]
It's okay, it seems like this is a rather uncommon problem.

What I tried so far was to remove the quarantine extended attribute from the file, which still resulted in the "not an OMNeT directory" error for `source setenv`. Since the last version that worked for me (before it mysteriously gave up) was 6.0.1, I tried installing that.

There, I once again get the following error, even though the file I unpacked was `omnetpp-6.0.1-macos-x86_64.tgz`:
```
-bash: [: too many arguments
-bash: [: too many arguments
-bash: [: too many arguments
Error: We are running on macOS x86_64, but you seem to have downloaded the Linux or Windows version of the installation package!
```
I checked the `setenv` file for what could cause this message, and it only happens when it cannot find `/tools/macos.x86_64` from this code:
```
elif [ -d $dir/tools/macos.x86_64 ]; then
      TOOLS_DIR=$dir/tools/macos.x86_64
``` 
I checked that, and that folder exists. In my `.profile` file under `/Users/juliasteiwer/.profile`, I only have this:

<img width="825" alt="Bildschirmfoto 2023-11-14 um 19 37 53" src="https://github.com/omnetpp/omnetpp/assets/42389995/7c180d2d-63bb-4f4f-937e-da1351da302a">

The reason I checked this was because I feel that something is preventing `$dir` from being executed properly for the OMNeT++ folder (which is `/Users/juliasteiwer/omnetpp-6.0.1` for me).

**Reply by JuliaSteiwer** [reply, marked as answer]
@rhornig I saw in the installation guide that for Apple Silicon (I still have Intel but irrelevant) OMNeT++ can be run under emulation.  So I thought, why not give this a try, called the command `arch -x86_64 /bin/zsh --login` in the terminal and then followed the normal installation procedure for OMNeT. This worked.

If I want to start OMNeT now, I always have to enter the emulation command `arch -x86_64 /bin/zsh --login` and then `source setenv` just to call `omnetpp`, but at least it works. Also, this uses zsh instead of bash, so that might have been an issue as well.

Thank you for taking your time to look at my issue!

**Reply by rhornig** [reply]
I dusted off an old iMacMini, but sadly it is so old that even downloading omnetpp from github was a giant task. I cannot reproduce this for now. If you figure out the exact cause of this issues sometime, I would love to know about it.
